src/bucket/BucketIndex.cpp

```diff
@@ -0,0 +1,244 @@
+// Copyright 2022 Stellar Development Foundation and contributors. Licensed
+// under the Apache License, Version 2.0. See the COPYING file at the root
+// of this distribution or at http://www.apache.org/licenses/LICENSE-2.0
+
+#include "bucket/BucketIndex.h"
+#include "bucket/Bucket.h"
+#include "bucket/LedgerCmp.h"
+#include "main/Config.h"
+#include "util/Fs.h"
+#include "util/Logging.h"
+#include "util/XDRStream.h"
+
+#include <Tracy.hpp>
+#include <fmt/chrono.h>
+#include <fmt/format.h>
+#include <ledger/LedgerHashUtils.h>
+
+namespace stellar
+{
+
+// Returns a poolshare trustline key with the given accountID and a PoolID
+// filled with the fill byte
+static LedgerKey
+getDummyPoolShareTrustlineKey(AccountID const& accountID, uint8_t fill)
+{
+    LedgerKey key(TRUSTLINE);
+    key.trustLine().accountID = accountID;
+    key.trustLine().asset.type(ASSET_TYPE_POOL_SHARE);
+    key.trustLine().asset.liquidityPoolID().fill(fill);
+    return key;
+}
+
+template <class IndexT>
+BucketIndexImpl<IndexT>::BucketIndexImpl(std::filesystem::path const& filename,
+                                         std::streamoff pageSize)
+    : mPageSize(pageSize)
+{
+    ZoneScoped;
+    if (filename.empty())
```

Comment:
Yes, this doesn't represent an invalid index, but an empty index. This happens when you index the entire BucketList and there are some empty initialized buckets. My thinking is every Bucket in the Bucket list should be indexed, even if the bucket is empty. It seemed to be a little cleaner to do this for the empty indexed buckets so that they pass the assert in `getIndex()` instead of putting in special cases for the empty initialized buckets in the BucketList.

Reply:
After working through this a little more, I think it does make more sense to throw. After a few changes all of the `BucketList` logic now only iterates through non-empty buckets, so indexing or getting the index of an empty bucket should be an error.